### ISSUE TITLE
Support playlists of DRM Protected Content with different DRM License Urls?

### ISSUE THREAD

**Issue by Jogan** [google/ExoPlayer#2394]
Is there any documentation for how to handle playing a list of drm protected content each with a dynamic license url (i.e. https://www.somewidevineurl.com/videos/{{video-id}}/)? 

Playlist:
- Video1 :  https://www.somewidevineurl.com/videos/1/
- Video2 :  https://www.somewidevineurl.com/videos/2/
- Video3 :  https://www.somewidevineurl.com/videos/3/

I am using the StreamingDrmSessionManager and then plan to use a ConcatenatingMediaSource for the playlist. The example uses a static license url for the HttpMediaDrmCallback.

**Comment by ojw28** [comment]
This is on our todo list, but is not something we currently support. Isn't the video id normally retrievable from the body of the key request? In which case one option would be to have a single license URL and have the server get the video id from the request body.

**Comment by Jogan** [comment]
I'll look into the server option, otherwise is the only way to release the player and then re-initialize it with the next item in the playlist? I guess I'll have to manually handle the playback for the playlist as well rather than using the ConcatenatingMediaSource.

**Comment by ojw28** [comment]
That's correct until we add support (we don't have an ETA, but we're aware of the fact we're not adequately supporting this use case).

**Comment by piatt** [comment]
Any updates on ETA for this? I also have the same need. Thanks!

**Comment by AquilesCanta** [comment]
This can now be solved by including the license server url in the manifest or, alternatively, injecting the server URL into the manifest through a custom playlist parser. We will keep working around per-MediaSource DRM so it's not a definitive solution, but it's a start.

**Comment by kushalgupta0565** [comment]
Hi,

I am new to exoplayer and these .MPD file things. But i am trying to get it. I am facing the similar challenge as mention above. I am trying to use `multiple DRM license` with a `single exoplayer instance`.

I have downloaded multiple license byte array using

```
public byte[] generateOfflineLicenseKeySet() {
        try {
            byte[] offlineLicenseKeySet = licenseComponent.getOfflineLicenseHelper().downloadLicense(getDrmInitData(licenseComponent.getDataSourceFactory()));
            saveOfflineLicenseKeySet(mpd_path, offlineLicenseKeySet);
            return offlineLicenseKeySet;
        } catch (DrmSession.DrmSessionException e) {
            e.printStackTrace();
        }
        return null;
    }
```

but not able to figure out how to pass them while creating the instance of Exoplayer. 

I am using `DashMediaSource.Factory` for generating the instance of MediaSource

```
new DashMediaSource.Factory(dataSourceFactory)
                    .setManifestParser(
                            new FilteringManifestParser<>(new DashManifestParser(), getOfflineStreamKeys(uri))).setTag(mediaData)
```

I tried to setMode for `DefaultDrmSessionManager` but it is working only for single `MediaSource`

  `drmSessionManager.setMode(DefaultDrmSessionManager.MODE_PLAYBACK,AppUtils.StringToByteArray(offlineLicenseKeyStr));`



Do i need to pass the `offlineLicenseKeySet `during the creation of MediaSource.
Please suggest me a way to use `multiple DRM license` with a `single exoplayer instance`

**Comment by kgrevehagen** [comment]
@AquilesCanta Any update on this issue or ETA? I guess the final solution also will resolve #5619? Any ideas on workarounds I can do in the meantime?

**Comment by AquilesCanta** [comment]
>  I guess the final solution also will resolve #5619?

Yes. 

> Any ideas on workarounds I can do in the meantime?

Sorry I can't provide a short guidance on this, changes are pretty big. Thanks for your patience.

**Comment by stefanmoro** [comment]
> This can now be solved by including the license server url in the manifest or, alternatively, injecting the server URL into the manifest through a custom playlist parser. We will keep working around per-MediaSource DRM so it's not a definitive solution, but it's a start.

@AquilesCanta Can you please give an example of where and what to insert into the (I assume) DASH manifest? If this works the way I understand it, it could probably be a workaround we can use until the refactoring you're working on is done.

Thanks
Stefan

**Comment by AquilesCanta** [comment]
@stefanmoro, see #3393. This feature should be available soon. Possibly in the next release.

**Comment by ojw28** [comment]
You can now set a `DrmSessionManager` directly on a `MediaSource`, as shown in the [demo app](https://github.com/google/ExoPlayer/blob/06f94815050b9f9e990c7c9d3d7b7eb0bbebecc1/demos/main/src/main/java/com/google/android/exoplayer2/demo/PlayerActivity.java#L501). Since this allows specifying a different `DrmSessionManager` on each `MediaSource`, it solves the problem that's being tracked by this issue.

This improvement will be included in the 2.11.0 release (we don't have an ETA yet).
